### PR TITLE
Only connect to new nodes on new cluster state

### LINKED ISSUE #29025: Do not attempt to re-establish connections when applying a cluster state update
Today, when a new cluster state is applied, we attempt to connect to all nodes in the new cluster state:

https://github.com/elastic/elasticsearch/blob/5904d936fabfa25b832b757461efa6b0d3e9acb9/server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java#L467

This set of nodes may include some nodes that were previously connected but which have failed, but whose failure is to be processed in a later cluster state update. Attempts to reconnect to these nodes will also fail, but may do so very slowly if they are unresponsive. On the other hand, the `NodesConnectionService` takes responsibility for periodically re-establishing connections that have dropped.

This means that there's no real need to attempt to re-establish these connections while applying a cluster state update, and #28920 is an example of a situation where it's undesirable to do so. Therefore it seems sensible to skip these nodes.

NB it's important that the nodes in the cluster state align with the nodes known to the `NodeConnectionsService`, so they can't simply be omitted in the call to `connectToNodes()` above.

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today, when a new cluster state is committed we attempt to connect to all of
its nodes as part of the application process. This is the right thing to do
with new nodes, and is a no-op on any already-connected nodes, but is
questionable on known nodes from which we are currently disconnected: there is
a risk that we are partitioned from these nodes so that any attempt to connect
to them will hang until it times out. This can dramatically slow down the
application of new cluster states which hinders the recovery of the cluster
during certain kinds of partition.

If nodes are disconnected from the master then it is likely that they are to be
removed as part of a subsequent cluster state update, so there's no need to try
and reconnect to them like this. Moreover there is no need to attempt to
reconnect to disconnected nodes as part of the cluster state application
process, because we periodically try and reconnect to any disconnected nodes,
and handle their disconnectedness gracefully in the meantime.

This commit alters this behaviour to avoid reconnecting to known nodes during
cluster state application.

Resolves #29025.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
I think this is a worthwhile improvement on the current situation, and it's a simple change, but I think it's worth revisiting in future since there's still room for improvement IMO: if the `ConnectionChecker` happens to be running when a partition is detected then it can block the cluster state update thread from getting hold of the `nodeLocks` that it needs for an extended period of time. Reducing the connection timeout (#29022) would help a bit here.

File: server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java [hunk 1]
```diff
@@ -83,8 +83,7 @@ public void connectToNodes(DiscoveryNodes discoveryNodes) {
         for (final DiscoveryNode node : discoveryNodes) {
             final boolean connected;
             try (Releasable ignored = nodeLocks.acquire(node)) {
-                nodes.putIfAbsent(node, 0);
-                connected = transportService.nodeConnected(node);
+                connected = (nodes.putIfAbsent(node, 0) != null);
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java]: can we avoid going on the management thread if we are truly connected? This will be relevant on the master where where the node is connected to during join validation. Also, I think the name `connected` for the variable is a bit mislead now. Maybe flip it and call it `shouldConnect`? 

It will also good to have a comment explaining the rational for this check.

PS - as discussed, it will be good to explore using this component on non-elected-master nodes which will make my condition tweak obsolete.

**Review by bleskes — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
@ywelsch as you predicted, `testAckedIndexing` failed during the validation phase after a number of runs. I pushed 9ea41666c4843c8c68eaf390174aa913eb9bf6ed.

**Comment by DaveCTurner** [comment]
@ywelsch I've run the full integ test suite overnight (162 times) on 8f977ff680e4ff28ad07dbaec6a69aaf41e9fa0f with no failures. WDYT?

File: server/src/test/java/org/elasticsearch/discovery/ClusterDisruptionIT.java [hunk 2]
```diff
@@ -207,7 +208,7 @@ public void testAckedIndexing() throws Exception {
                                 assertTrue("doc [" + id + "] indexed via node [" + ackedDocs.get(id) + "] not found",
                                     client(node).prepareGet("test", "type", id).setPreference("_local").get().isExists());
                             }
-                        } catch (AssertionError | NoShardAvailableActionException e) {
+                        } catch (AssertionError | NoShardAvailableActionException | NodeNotConnectedException e) {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/discovery/ClusterDisruptionIT.java]: Is this change still required?

**Review by ywelsch — COMMENTED** [review]
I'm a bit on the fence for this change. Yes, it will help in the situation described, but I wonder what adversarial effects it will have after a network partition (We work around these adversarial effects in our tests by explicitly reconnecting the nodes). I wonder if we should force `reconnectToKnownNodes` whenever the node applies a cluster state from a fresh master. Yes, this will slow down publishing a little, but only the first cluster state that's received from the new master, not subsequent ones.

**Comment by DaveCTurner** [comment]
I've decided to close this. It mostly helps, but with this approach there's still a chance of blocking a cluster state update on a connection to a blackhole because of the node locks, and I think it'd be better for future debugging/analysis if this wasn't based on luck.
